### PR TITLE
Add support for MAIN_MODULE/SIDE_MODULE with the wasm backend

### PR DESCRIPTION
**Pull request by sbc100**
A few tests are still disabled, but mostly for reasonable reasons.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1090,8 +1086,19 @@ def check(input_file):
         'allocate',
         'getMemory',
       ]
+      if shared.Settings.WASM_BACKEND:
+        shared.Settings.EXPORTED_RUNTIME_METHODS += [
```
Comment (kripken) [inline emcc.py]: is this always needed? or just for linkable modules?

File: emscripten.py [hunk 2]
```diff
@@ -687,6 +687,8 @@ def read_proxied_function_signatures(asmConsts):
   if shared.Settings.WASM_BACKEND:
     shared.Settings.WASM_TABLE_SIZE = metadata['tableSize']
     shared.Settings.BINARYEN_FEATURES = metadata['features']
+    if shared.Settings.MAIN_MODULE:
+      shared.Settings.WASM_TABLE_SIZE += 1
```
Comment (kripken) [inline emscripten.py]: this looks surprising, why?

File: emscripten.py [hunk 3]
```diff
@@ -2213,8 +2251,11 @@ def debug_copy(src, dst):
     shared.check_call(sourcemap_cmd)
     debug_copy(base_source_map, 'base_wasm.map')
 
-  cmd = [wasm_emscripten_finalize, base_wasm, '-o', wasm,
-         '--global-base=%s' % shared.Settings.GLOBAL_BASE]
+  cmd = [wasm_emscripten_finalize, base_wasm, '-o', wasm]
+  if shared.Settings.RELOCATABLE:
+    cmd.append('--global-base=0')
```
Comment (kripken) [inline emscripten.py]: is this just ignored for relocatable modules? or is the value 0 somehow important?

File: emscripten.py [hunk 4]
```diff
@@ -2310,7 +2353,7 @@ def create_sending_wasm(invoke_funcs, forwarded_json, metadata):
   if shared.Settings.SAFE_HEAP:
     basic_funcs += ['segfault', 'alignfault']
 
-  basic_vars = ['DYNAMICTOP_PTR']
+  basic_vars = ['DYNAMICTOP_PTR', 'tempDoublePtr']
```
Comment (kripken) [inline emscripten.py]: why is that needed? for wasm hopefully we don't need it at all, so I'm surprised to see it here.

File: src/jsifier.js [hunk 5]
```diff
@@ -160,11 +160,29 @@ function JSify(data, functionsOnly) {
           // emit a stub that will fail at runtime
           LibraryManager.library[ident] = new Function("err('missing function: " + ident + "'); abort(-1);");
         } else {
-          var target = (MAIN_MODULE ? '' : 'parent') + "Module['_" + ident + "']";
+          var isGlobalAccessor = ident.startsWith('g$');
+          var isFunctionAddressAccessor = ident.startsWith('fp$');
+          //if (isFunctionAddressAccessor)
```
Comment (kripken) [inline src/jsifier.js]: temp comments?

File: src/jsifier.js [hunk 6]
```diff
@@ -160,11 +160,29 @@ function JSify(data, functionsOnly) {
...
+            //return;
+          var realIdent = ident;
+          if (isGlobalAccessor)
```
Comment (kripken) [inline src/jsifier.js]: `{ }` around if body

File: src/library_pthread.js [hunk 7]
```diff
@@ -211,7 +211,7 @@ var LibraryPThread = {
       PThread.pthreads = {};
       for (var t in PThread.unusedWorkerPool) {
         var pthread = PThread.unusedWorkerPool[t];
-        if (pthread) {
+        if (pthread) {cancelDone
```
Comment (kripken) [inline src/library_pthread.js]: this doesn't look intentional

File: src/preamble.js [hunk 8]
```diff
@@ -1032,7 +1032,11 @@ Module['asm'] = function(global, env, providedBuffer) {
     'element': 'anyfunc'
   });
   env['__memory_base'] = {{{ GLOBAL_BASE }}}; // tell the memory segments where to place themselves
+#if WASM_BACKEND
+  env['__table_base'] = 1; // table starts at 1 by default (even in dynamic linking, for the main module)
```
Comment (kripken) [inline src/preamble.js]: same comment on this and the line below. are both valid? why is this 1?

File: tools/cache.py [hunk 9]
```diff
@@ -45,10 +45,14 @@ def try_remove_ending(thestring, ending):
...
-        dirname = os.path.join(dirname, 'wasm_o')
       else:
-        dirname = os.path.join(dirname, 'wasm_bc')
+        dirname = os.path.join(dirname, 'wasm')
+        if shared.Settings.WASM_OBJECT_FILES:
+          dirname = os.path.join(dirname, 'obj')
+        else:
+          dirname = os.path.join(dirname, 'bc')
+        if shared.Settings.RELOCATABLE:
+          dirname = os.path.join(dirname, 'pic')
```
Comment (kripken) [inline tools/cache.py]: do we need `pic-obj`, `pic-bc`?

File: tools/shared.py [hunk 10]
```diff
@@ -907,7 +907,7 @@ def __init__(self, environ=os.environ):
 
     self.CANONICAL_TEMP_DIR = get_canonical_temp_dir(self.TEMP_DIR)
 
-    if DEBUG:
+    if DEBUG or os.getenv('EMCC_SAVE_DIR'):
```
Comment (kripken) [inline tools/shared.py]: looks like a separate change from the rest of the PR?

File: tools/cache.py (hunk 9 above)
Comment (sbc100) [reply tools/cache.py]: This code already takes care of that by just adding another level of nesting.  However I decided to flatten it out.

File: emscripten.py (hunk 3 above)
Comment (sbc100) [reply emscripten.py]: Added comment.

**Comment by sbc100** [comment]
Thanks for all the great feedback.  Should be a much cleaner change now.

File: emscripten.py [hunk 11]
```diff
@@ -2228,7 +2268,15 @@ def debug_copy(src, dst):
...
   if not shared.Settings.MEM_INIT_IN_WASM:
     cmd.append('--separate-data-segments=' + memfile)
-  if not shared.Settings.SIDE_MODULE:
+  if shared.Settings.SIDE_MODULE:
+    cmd.append('--side-module')
+  else:
+    # Global base is used to calulate the size of the static data segment.  The
+    # value should match the --global-base passed to lld in Building.link_lld()
+    if shared.Settings.RELOCATABLE:
+      cmd.append('--global-base=0')
```
Comment (kripken) [inline emscripten.py]: the comment doesn't explain why it's 0 in the relocatable case? (and I still don't understand why myself too :)

File: src/jsifier.js [hunk 12]
```diff
@@ -156,15 +156,33 @@ function JSify(data, functionsOnly) {
...
+            var what = 'function';
+            if (isGlobalAccessor) {
+              what = 'global';
+            }
+            assertion += 'if (!' + target + ') abort("external ' + what + ' \'' + realIdent + '\' is missing. perhaps a side module was not linked in? if this function was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment");\n';
+
+          }
+          var functionBody;
+          if (isGlobalAccessor) {
+            functionBody = assertion + "return " + target + ";"
```
Comment (kripken) [inline src/jsifier.js]: a global accessor is still a function that needs to be called, isn't it? I'm surprised not to see a call here (just `()`) like 2 lines down?

File: src/preamble.js [hunk 13]
```diff
@@ -1032,7 +1032,14 @@ Module['asm'] = function(global, env, providedBuffer) {
     'element': 'anyfunc'
   });
   env['__memory_base'] = {{{ GLOBAL_BASE }}}; // tell the memory segments where to place themselves
+#if WASM_BACKEND
+  // We reserve slot 0 in the table for the NULL function pointer.
+  // This means the __table_base for the main module (even in dynamic linking)
+  // is alwasy 1.
```
Comment (kripken) [inline src/preamble.js]: typo in `always`

File: src/preamble.js [hunk 14]
```diff
@@ -1032,7 +1032,14 @@ Module['asm'] = function(global, env, providedBuffer) {
...
+  env['__table_base'] = 1;
```
Comment (kripken) [inline src/preamble.js]: this changes the non-relocatable case - do our non-relocatable wasm modules not depend on that value for placing their table segments? it seems like this would move them forward by 1, unless we never use this param in the non-relocatable case?

File: tests/test_other.py [hunk 15]
```diff
@@ -1305,6 +1305,10 @@ def tmp_source(name, code):
     run_process([PYTHON, EMCC] + args + libs)
     self.assertContained('result: 42', run_js('a2.out.js'))
 
+  # The fastcomp path will deliberatly ignore duplicate input file in order
```
Comment (kripken) [inline tests/test_other.py]: typo in `deliberately`

Comment (kripken) [reply tests/test_other.py]: also `file => files`

File: tests/test_other.py [hunk 16]
```diff
@@ -1305,6 +1305,10 @@ def tmp_source(name, code):
...
+  # to allow "libA.so" on the command line twice. The is not realy .so support
```
Comment (kripken) [inline tests/test_other.py]: typo in `really`

File: tests/test_other.py [hunk 17]
```diff
@@ -1305,6 +1305,10 @@ def tmp_source(name, code):
...
+  # and the .so files are reallly bitcode.
```
Comment (kripken) [inline tests/test_other.py]: do native linkers not allow linking in the same library twice (with no effects, same as if it were linked once)? or is that just for `-l` but not for providing an `.so` file?

File: tests/test_other.py [hunk 18]
```diff
@@ -1323,7 +1327,7 @@ def test_redundant_link(self):
 
     Building.emcc(lib_name, output_filename='libA.so')
 
-    Building.emcc(main_name, ['libA.so'] * 2, output_filename='a.out.js')
+    Building.emcc(main_name, ['libA.so', 'libA.so'], output_filename='a.out.js')
```
Comment (kripken) [inline tests/test_other.py]: heh, i don't care either way, but why didn't you like the `* 2` here? :)

File: tests/test_other.py [hunk 19]
```diff
@@ -3840,6 +3844,8 @@ def test_side_module_without_proper_target(self):
     # SIDE_MODULE is only meaningful when compiling to wasm (or js+wasm)
     # otherwise, we are just linking bitcode, and should show an error
     for wasm in [0, 1]:
+      if self.is_wasm_backend():
+        continue
```
Comment (kripken) [inline tests/test_other.py]: this test now has both `@no_wasm_backend` and this line here

File: tests/test_other.py (hunk 18 above)
Comment (sbc100) [reply tests/test_other.py]: I'll revert that.  I just though it was too clever for its own good.. especially in test code :)

File: tests/test_other.py (hunk 17 above)
Comment (sbc100) [reply tests/test_other.py]: They may for .so files, because of the native of .so files.  And for .a files for a different reason.

But we don't support .so files in that sense.   Our support for .so files should be removed IMHO since its confusing and doesn't do what people would want from .so file support.

File: src/preamble.js (hunk 14 above)
Comment (sbc100) [reply src/preamble.js]: __table_base and __memory_base (at least under the wasm backend) are ignored except for relocatable code.

I added a comment to clarify.

File: src/jsifier.js (hunk 12 above)
Comment (sbc100) [reply src/jsifier.js]: For each global address in the system there exist two different accessors:

```
Module['g$foo'] holds a function that returns Module['foo']

Module['foo'] holds that actual address
```

Modules export addresses in the form of foo `foo` but import in the form of `g$foo`.   The `g$foo` functions are created by the JS glue.
